### ISSUE TITLE
NullPointerException using @JsonProperty to rename a MongoDB property

### ISSUE THREAD

**Issue by mjustin** [spring-projects/spring-data-rest#2037]
I have a Spring Boot app using Spring Data MongoDB with Spring Data REST.  It uses the [`@JsonProperty`](http://fasterxml.github.io/jackson-annotations/javadoc/2.12/com/fasterxml/jackson/annotation/JsonProperty.html) annotation to rename some properties, including one in a nested object.

I recently tried upgrading Spring Boot from 2.4.5 to 2.5.2 (upgrading Spring Data from 3.4.8 to 3.5.1).  With this change, create REST requests for that document that previously succeeded now fail.  Specifically, it occurs when I have a nested object that is renaming a JSON property using `@JsonProperty`.  It does not fail if the `@JsonProperty` is removed, or if the `@JsonProperty` rename is on the main document itself and not a nested object.

The relevant portion of the stack trace is:

```
java.lang.NullPointerException: null
	at org.springframework.data.rest.webmvc.json.AggregateReferenceResolvingModule$AggregateReferenceDeserializerModifier.updateBuilder(AggregateReferenceResolvingModule.java:109) ~[spring-data-rest-webmvc-3.5.2.jar:3.5.2]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.buildBeanDeserializer(BeanDeserializerFactory.java:287) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.createBeanDeserializer(BeanDeserializerFactory.java:150) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer2(DeserializerCache.java:414) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer(DeserializerCache.java:349) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:264) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.DeserializationContext.findNonContextualValueDeserializer(DeserializationContext.java:581) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.resolve(BeanDeserializerBase.java:539) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:293) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer(DeserializationContext.java:591) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.ObjectMapper._findRootDeserializer(ObjectMapper.java:4733) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4594) ~[jackson-databind-2.12.3.jar:2.12.3]
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3601) ~[jackson-databind-2.12.3.jar:2.12.3]
	at org.springframework.http.converter.json.AbstractJackson2HttpMessageConverter.readJavaType(AbstractJackson2HttpMessageConverter.java:378) ~[spring-web-5.3.8.jar:5.3.8]
	at org.springframework.http.converter.json.AbstractJackson2HttpMessageConverter.readInternal(AbstractJackson2HttpMessageConverter.java:350) ~[spring-web-5.3.8.jar:5.3.8]
	at org.springframework.http.converter.AbstractHttpMessageConverter.read(AbstractHttpMessageConverter.java:199) ~[spring-web-5.3.8.jar:5.3.8]
	at org.springframework.data.rest.webmvc.config.PersistentEntityResourceHandlerMethodArgumentResolver.read(PersistentEntityResourceHandlerMethodArgumentResolver.java:246) ~[spring-data-rest-webmvc-3.5.2.jar:3.5.2]
	at org.springframework.data.rest.webmvc.config.PersistentEntityResourceHandlerMethodArgumentResolver.lambda$read$6(PersistentEntityResourceHandlerMethodArgumentResolver.java:202) ~[spring-data-rest-webmvc-3.5.2.jar:3.5.2]
	at java.base/java.util.Optional.orElseGet(Optional.java:369) ~[na:na]
	at org.springframework.data.rest.webmvc.config.PersistentEntityResourceHandlerMethodArgumentResolver.read(PersistentEntityResourceHandlerMethodArgumentResolver.java:202) ~[spring-data-rest-webmvc-3.5.2.jar:3.5.2]
	at org.springframework.data.rest.webmvc.config.PersistentEntityResourceHandlerMethodArgumentResolver.resolveArgument(PersistentEntityResourceHandlerMethodArgumentResolver.java:132) ~[spring-data-rest-webmvc-3.5.2.jar:3.5.2]
	at org.springframework.web.method.support.HandlerMethodArgumentResolverComposite.resolveArgument(HandlerMethodArgumentResolverComposite.java:121) ~[spring-web-5.3.8.jar:5.3.8]
	at org.springframework.web.method.support.InvocableHandlerMethod.getMethodArgumentValues(InvocableHandlerMethod.java:170) ~[spring-web-5.3.8.jar:5.3.8]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:137) ~[spring-web-5.3.8.jar:5.3.8]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:106) ~[spring-webmvc-5.3.8.jar:5.3.8]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:894) ~[spring-webmvc-5.3.8.jar:5.3.8]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808) ~[spring-webmvc-5.3.8.jar:5.3.8]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-5.3.8.jar:5.3.8]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1063) ~[spring-webmvc-5.3.8.jar:5.3.8]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963) ~[spring-webmvc-5.3.8.jar:5.3.8]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) ~[spring-webmvc-5.3.8.jar:5.3.8]
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909) ~[spring-webmvc-5.3.8.jar:5.3.8]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:681) ~[tomcat-embed-core-9.0.48.jar:4.0.FR]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) ~[spring-webmvc-5.3.8.jar:5.3.8]
	at org.springframework.test.web.servlet.TestDispatcherServlet.service(TestDispatcherServlet.java:72) ~[spring-test-5.3.8.jar:5.3.8]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:764) ~[tomcat-embed-core-9.0.48.jar:4.0.FR]
	at org.springframework.mock.web.MockFilterChain$ServletFilterProxy.doFilter(MockFilterChain.java:167) ~[spring-test-5.3.8.jar:5.3.8]
	at org.springframework.mock.web.MockFilterChain.doFilter(MockFilterChain.java:134) ~[spring-test-5.3.8.jar:5.3.8]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-5.3.8.jar:5.3.8]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.3.8.jar:5.3.8]
	at org.springframework.mock.web.MockFilterChain.doFilter(MockFilterChain.java:134) ~[spring-test-5.3.8.jar:5.3.8]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-5.3.8.jar:5.3.8]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.3.8.jar:5.3.8]
	at org.springframework.mock.web.MockFilterChain.doFilter(MockFilterChain.java:134) ~[spring-test-5.3.8.jar:5.3.8]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-5.3.8.jar:5.3.8]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.3.8.jar:5.3.8]
	at org.springframework.mock.web.MockFilterChain.doFilter(MockFilterChain.java:134) ~[spring-test-5.3.8.jar:5.3.8]
	at org.springframework.test.web.servlet.MockMvc.perform(MockMvc.java:183) ~[spring-test-5.3.8.jar:5.3.8]
```

It's seems like this might be related to/duplicating other 3.5 issues reported elsewhere, such as #2023

## Example

**`build.gradle`**

```groovy
plugins {
    id 'io.spring.dependency-management' version '1.0.11.RELEASE'
    id 'java'
    id 'org.springframework.boot' version '2.5.2' //Works with 2.4.5
}

sourceCompatibility = '11'

repositories {
    mavenCentral()
}

dependencies {
    implementation 'org.springframework.boot:spring-boot-starter-data-rest'
    implementation 'org.springframework.boot:spring-boot-starter-data-mongodb'

    testImplementation 'org.springframework.boot:spring-boot-starter-test'
    testImplementation 'com.playtika.testcontainers:embedded-mongodb:2.0.7'
    testImplementation('org.springframework.cloud:spring-cloud-starter-bootstrap:3.0.3') {
        because 'required by com.playtika.testcontainers:embedded-mongodb'
    }
}

test {
    useJUnitPlatform()
}
```

```java
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

@Document
public class MyDocument {
    @Id
    private String id;

    private NestedType value;

    public MyDocument() { }

    public MyDocument(String id, String value) {
        this.id = id;
        this.value = new NestedType(value);
    }

    public String getId() {return id;}
    public void setId(String id) {this.id = id;}
    public NestedType getValue() {return value;}
    public void setValue(NestedType value) {this.value = value;}
}
```

```java
import com.fasterxml.jackson.annotation.JsonProperty;
import org.springframework.data.mongodb.core.mapping.Field;

public class NestedType {
    @Field("renamedValue")
    @JsonProperty("renamedValue")
    private String value;

    public NestedType() {}
    public NestedType(String value) { this.value = value; }
    public String getValue() {return value;}
    public void setValue(String value) {this.value = value;}
}
```

```java
import org.springframework.data.repository.PagingAndSortingRepository;

public interface MyDocumentRepository extends PagingAndSortingRepository<MyDocument, String> {}
```

```java
import com.fasterxml.jackson.databind.json.JsonMapper;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.web.servlet.MockMvc;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
@TestPropertySource(properties = {
        "embedded.containers.enabled=true",
        "spring.data.mongodb.uri=mongodb://${embedded.mongodb.host}:${embedded.mongodb.port}/${embedded.mongodb.database}"
})
public class RestTest {
    @Autowired
    private MockMvc mockMvc;

    @Test
    public void restPost() throws Exception {
        MyDocument document = new MyDocument("my-id", "My Value");

        mockMvc
                .perform(post("/myDocuments")
                        .contentType("application/json")
                        .content(JsonMapper.builder().build().writeValueAsString(document)))
                .andExpect(status().isCreated());
    }
}
```

**Comment by mjustin** [comment]
I've created a workaround to allow the application to continue to function as before, by skipping the registration of the `AggregateReferenceResolvingModule` in the `ObjectMapper`:

```java
/**
 * Workaround for https://github.com/spring-projects/spring-data-rest/issues/2037.  Identical to
 * {@link RepositoryRestMvcConfiguration}, except that it does not register an
 * {@link org.springframework.data.rest.webmvc.json.AggregateReferenceResolvingModule} with the object mapper.
 *
 * @see RepositoryRestMvcConfiguration
 * @see <a href="https://github.com/spring-projects/spring-data-rest/issues/2037">spring-data-rest#2037</a>
 */
@Component
public class CopyWriteApiRepositoryRestConfiguration extends RepositoryRestMvcConfiguration {
    // Private fields in RepositoryRestMvcConfiguration
    private final ObjectProvider<GeoModule> geoModule;

    private final ObjectProvider<ObjectMapper> objectMapper;

    // Other private fields
    private final EnumTranslator enumTranslator;

    private final RepositoryRestConfiguration repositoryRestConfiguration;

    private final List<RepositoryRestConfigurer> repositoryRestConfigurers;

    public CopyWriteApiRepositoryRestConfiguration(
            ApplicationContext context,
            ObjectFactory<ConversionService> conversionService,
            ObjectProvider<LinkRelationProvider> relProvider,
            ObjectProvider<CurieProvider> curieProvider,
            ObjectProvider<HalConfiguration> halConfiguration,
            ObjectProvider<ObjectMapper> objectMapper,
            ObjectProvider<RepresentationModelProcessorInvoker> invoker,
            ObjectProvider<MessageResolver> resolver,
            ObjectProvider<GeoModule> geoModule,
            ObjectProvider<PathPatternParser> parser,
            @Lazy EnumTranslator enumTranslator,
            @Lazy RepositoryRestConfiguration repositoryRestConfiguration,
            List<RepositoryRestConfigurer> repositoryRestConfigurers) {
        super(context, conversionService, relProvider, curieProvider, halConfiguration, objectMapper, invoker, resolver,
                geoModule, parser);

        this.geoModule = geoModule;
        this.objectMapper = objectMapper;
        this.enumTranslator = enumTranslator;
        this.repositoryRestConfiguration = repositoryRestConfiguration;
        this.repositoryRestConfigurers = repositoryRestConfigurers;
    }

    /**
     * This is intended to be identical in function to the superclass method, excep that it does not register an
     * {@link org.springframework.data.rest.webmvc.json.AggregateReferenceResolvingModule}.
     */
    @Override
    protected ObjectMapper basicObjectMapper() {
        ObjectMapper mapper = this.objectMapper.getIfAvailable();
        ObjectMapper objectMapper = mapper == null ? new ObjectMapper() : mapper.copy();

        objectMapper.configure(SerializationFeature.INDENT_OUTPUT, true);
        objectMapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

        repositoryRestConfigurers.forEach(c -> c.configureJacksonObjectMapper(objectMapper));

        objectMapper.registerModule(geoModule.getObject());

        // Do not register the AggregateReferenceResolvingModule to work around the issue
//        objectMapper.registerModule(new AggregateReferenceResolvingModule(
//                new UriToEntityConverter(persistentEntities.get(), repositoryInvokerFactory.get(), repositories.get()),
//                resourceMappings.get()));

        if (repositoryRestConfiguration.isEnableEnumTranslation()) {
            objectMapper.registerModule(new JacksonSerializers(enumTranslator));
        }

        Jackson2DatatypeHelper.configureObjectMapper(objectMapper);

        return objectMapper;
    }
}
```

**Comment by schauder** [comment]
This is a duplicate of #2033

**Comment by mjustin** [comment]
Confirmed that this is fixed in 3.5.3-SNAPSHOT.
